File: iOS/G3MiOSSDK/Specific/SQDatabase.m
//
//  SQDatabase.m
//  wikiglob3-ios
//
//  Created by Diego Gomez Deck on 08/03/12.
//  Copyright (c) 2012 IGO Software SL. All rights reserved.
//

#import "SQDatabase.h"

@implementation SQResultSet{
  sqlite3_stmt *_stmt;
}

@synthesize queryParamCount = _queryParamCount;
@synthesize columnCount = _columnCount;
@synthesize sql = _sql;
@synthesize db = _db;

+ (SQResultSet*) resultSetWithDB: (SQDatabase*) db
                           query: (NSString*)   sql
                       arguments: (NSArray*)    args
{
  return [[SQResultSet alloc] initWithWithDB: (SQDatabase*) db
                                       query: sql
                                   arguments: args];
}

- (id) initWithWithDB: (SQDatabase*) db
                query: (NSString*)   sql
            arguments: (NSArray*)    args;
{
  self = [super init];
  if (self) {
    _db = db;
    _stmt = NULL;
    if (![db prepareSql: sql
            inStatement: &_stmt]) {
      NSLog(@"Can't prepare statement \"%@\", error=(%i:%@)",
            sql,
            [_db errorCode],
            [_db errorMessage]);
      return nil;
    }

    _queryParamCount = sqlite3_bind_parameter_count(_stmt);
    //NSLog(@"queryParamCount=%i", _queryParamCount);

    int i = 0;
    while (i++ < _queryParamCount) {
      [_db bindObject: [args objectAtIndex:(NSUInteger)(i - 1)]
             toColumn: i
          inStatement: _stmt];
    }

    _columnCount = sqlite3_column_count(_stmt);
    //NSLog(@"columnCount=%i", _columnCount);
  }
  return self;
}

- (void) close
{
  if (_stmt) {
    //    if (!sqlite3_finalize(_stmt)) {
    //      NSLog(@"Can't finalize ResulSet \"%@\"", _sql);
    //    }
    sqlite3_finalize(_stmt);

    _stmt = NULL;
  }
}

- (BOOL) next
{
  BOOL result = [_db hasData: _stmt];
  if (!result) {
    //NSLog(@"- Auto closing ResultSet");
    [self close];
  }
  return result;
}

-(void) dealloc
{
  [self close];
}

- (BOOL) isNullColumnByIndex: (NSInteger) index
{
  int columnType = sqlite3_column_type(_stmt, index);
	return (columnType == SQLITE_NULL);
}

- (NSInteger) integerColumnByIndex: (NSInteger) index
{
  int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return 0;
  }

  return sqlite3_column_int(_stmt, index);
}

- (NSString*) stringColumnByIndex: (NSInteger) index
{
  int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return nil;
  }

  return [NSString stringWithUTF8String:(const char* ) sqlite3_column_text(_stmt, index)];
}

- (NSData*) dataColumnByIndex: (NSInteger) index
{
  NSInteger blobLength = sqlite3_column_bytes(_stmt, index);
  return [NSData dataWithBytes: sqlite3_column_blob(_stmt, index)
                        length: (NSUInteger)blobLength];
}

- (double) doubleColumnByIndex: (NSInteger) index
{
  int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return 0;
  }

  return sqlite3_column_double(_stmt, index);
}

@end



@implementation SQDatabase

@synthesize dbPath           = _dbPath;
@synthesize busyRetryTimeout = _busyRetryTimeout;


+ (SQDatabase*) databaseWithPath: (NSString *) dbPath
{
  return [[SQDatabase alloc] initWithPath: dbPath];
}

- (id) initWithPath: (NSString*) dbPath
{
  self = [super init];
  if (self) {
    _db               = NULL;
    _dbPath           = dbPath;
    _busyRetryTimeout = 10;
  }
  return self;
}

- (BOOL) isOpen
{
  if (_db)
    return YES;
  else
    return NO;
}

- (BOOL) openReadWrite
{
  [self close];

  const char* dbpathC = [self.dbPath UTF8String];

  if (sqlite3_open_v2 (dbpathC, &_db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, NULL) == SQLITE_OK) {
    return YES;
  }
  else {
    _db = NULL;

    NSLog(@"Can't open readwrite database \"%@\"", self.dbPath);
    return NO;
  }
}

- (BOOL) openReadOnly
{
  [self close];

  const char* dbpathC = [self.dbPath UTF8String];

  if (sqlite3_open_v2(dbpathC, &_db, SQLITE_OPEN_READONLY, NULL) == SQLITE_OK) {
    return YES;
  }
  else {
    _db = NULL;

    NSLog(@"Can't open readonly database \"%@\"", self.dbPath);
    return NO;
  }
}


- (void) close
{
  if ([self isOpen]) {
    //    if (sqlite3_close(_db)) {
    //      _db = NULL;
    //    }
    //    else {
    //      NSLog(@"Can't close database \"%@\"", self.dbPath);
    //    }
    sqlite3_close(_db);
    _db = NULL;
  }
}

- (BOOL) prepareSql: (NSString*)      sql
        inStatement: (sqlite3_stmt**) stmt
{
	int numOfRetries = 0;

	do {
		int rc = sqlite3_prepare_v2(_db, [sql UTF8String], -1, stmt, NULL);
		if (rc == SQLITE_OK) {
			return YES;
    }

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				NSLog(@"SQLite Busy 1: %@", _dbPath);
				break;
			}

      usleep(10000 /* 10ms */);
		}
    else {
			NSLog(@"SQLite Prepare Failed: %s", sqlite3_errmsg(_db));
			NSLog(@" - Query: %@", sql);
			break;
		}
	}
  while (numOfRetries++ <= _busyRetryTimeout);

	return NO;
}

- (SQResultSet*) executeQuery: (NSString *) sql, ...
{
  va_list args;
	va_start(args, sql);

	NSMutableArray *argsArray = [[NSMutableArray alloc] init];
	NSUInteger i;
	for (i = 0; i < [sql length]; ++i) {
		if ([sql characterAtIndex:i] == '?')
			[argsArray addObject:va_arg(args, id)];
	}

	va_end(args);

	return [self executeQuery: sql
                  arguments: argsArray];
}

- (SQResultSet*) executeQuery: (NSString*) sql
                    arguments: (NSArray*)  args
{
  return [SQResultSet resultSetWithDB: self
                                query: sql
                            arguments: args];
}


- (BOOL) hasData:( sqlite3_stmt*) stmt
{
	int numOfRetries = 0;

	do {
    //    if (numOfRetries > 0) {
    //      NSLog(@"SQLite Busy, step %i: %@", numOfRetries, _dbPath);
    //    }

		int rc = sqlite3_step(stmt);

		if (rc == SQLITE_ROW) {
			return YES;
    }

		if (rc == SQLITE_DONE) {
			break;
    }

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				NSLog(@"SQLite Busy 2: %@", _dbPath);
				break;
			}

      usleep(10000 /* 10ms */);
		}
    else {
			NSLog(@"SQLite Prepare Failed: %s", sqlite3_errmsg(_db));
			break;
		}
	}
  while (numOfRetries++ <= _busyRetryTimeout);

	return NO;
}

- (NSInteger) errorCode
{
	return sqlite3_errcode(_db);
}

- (NSString*) errorMessage
{
	return [NSString stringWithFormat:@"%s", sqlite3_errmsg(_db)];
}

- (void) bindObject: (id)            obj
           toColumn: (int)           idx
        inStatement: (sqlite3_stmt*) stmt
{
	if (obj == nil || obj == [NSNull null]) {
		sqlite3_bind_null(stmt, idx);
	}
  else if ([obj isKindOfClass:[NSData class]]) {
		sqlite3_bind_blob(stmt, idx, [obj bytes], (int) [obj length], SQLITE_STATIC);
	}
  else if ([obj isKindOfClass:[NSDate class]]) {
		sqlite3_bind_double(stmt, idx, [obj timeIntervalSince1970]);
	}
  else if ([obj isKindOfClass:[NSNumber class]]) {
		if (!strcmp([obj objCType], @encode(BOOL))) {
			sqlite3_bind_int(stmt, idx, [obj boolValue] ? 1 : 0);
		}
    else if (!strcmp([obj objCType], @encode(int))) {
			sqlite3_bind_int64(stmt, idx, [obj longValue]);
		}
    else if (!strcmp([obj objCType], @encode(long))) {
			sqlite3_bind_int64(stmt, idx, [obj longValue]);
		}
    else if (!strcmp([obj objCType], @encode(float))) {
			sqlite3_bind_double(stmt, idx, [obj floatValue]);
		}
    else if (!strcmp([obj objCType], @encode(double))) {
			sqlite3_bind_double(stmt, idx, [obj doubleValue]);
		}
    else {
			sqlite3_bind_text(stmt, idx, [[obj description] UTF8String], -1, SQLITE_STATIC);
		}
	}
  else {
		sqlite3_bind_text(stmt, idx, [[obj description] UTF8String], -1, SQLITE_STATIC);
	}
}

- (BOOL) executeStatament: (sqlite3_stmt *) stmt
{
	int numOfRetries = 0;
	int rc;

	do {
		rc = sqlite3_step(stmt);
		if (rc == SQLITE_OK || rc == SQLITE_DONE)
			return YES;

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				NSLog(@"SQLite Busy 3: %@", _dbPath);
				break;
			}

      usleep(10000 /* 10ms */);
		}
    else {
			NSLog(@"SQLite Step Failed: %s", sqlite3_errmsg(_db));
			break;
		}
	}
  while (numOfRetries++ <= _busyRetryTimeout);

	return NO;
}

- (BOOL) executeNonQuery: (NSString*) sql, ...
{
	va_list args;
	va_start(args, sql);

	NSMutableArray *argsArray = [[NSMutableArray alloc] init];
	NSUInteger i;
	for (i = 0; i < [sql length]; ++i) {
		if ([sql characterAtIndex:i] == '?')
			[argsArray addObject:va_arg(args, id)];
	}

	va_end(args);

	BOOL success = [self executeNonQuery:sql arguments:argsArray];

	return success;
}

- (BOOL) executeNonQuery: (NSString*) sql
               arguments: (NSArray*)  args
{
	sqlite3_stmt *sqlStmt;

	if (![self prepareSql:sql inStatement:(&sqlStmt)]) {
		return NO;
  }

	int i = 0;
	int queryParamCount = sqlite3_bind_parameter_count(sqlStmt);
	while (i++ < queryParamCount) {
		[self bindObject:[args objectAtIndex:(NSUInteger) (i - 1)] toColumn:i inStatement:sqlStmt];
  }

	BOOL success = [self executeStatament:sqlStmt];

	sqlite3_finalize(sqlStmt);
	return success;
}

- (BOOL) commit
{
	return [self executeNonQuery:@"COMMIT TRANSACTION;"];
}

- (BOOL) rollback
{
	return [self executeNonQuery:@"ROLLBACK TRANSACTION;"];
}

- (BOOL) beginTransaction
{
	return [self executeNonQuery:@"BEGIN EXCLUSIVE TRANSACTION;"];
}

- (BOOL) beginDeferredTransaction
{
	return [self executeNonQuery:@"BEGIN DEFERRED TRANSACTION;"];
}

@end

File: Windows/G3MWindowsSDK/Specific/SQDatabase.cpp

#include "SQDatabase.hpp"
#include "ILogger.hpp"
#include "IStringBuilder.hpp"
#include <thread>
#include <ctime>
#include <sstream>
#include <cstdarg>
//#include <stdarg.h>



ContentValue::ContentValue(unsigned char* content, int size, std::string type){

	_content = content;
	_size = size;
	_type = type;
}

//ContentValue::~ContentValue(){
//
//}

//----------------------------------------------------------------------------------



SQResultSet* SQResultSet::initForDatabase(const SQDatabase* db, std::string* query, std::vector<ContentValue*> args){

	try{
		return new SQResultSet(db, query, args);
	}
	catch (std::string e){
		ILogger::instance()->logError(e.c_str());
	}
	return NULL;
}

SQResultSet::SQResultSet(const SQDatabase* db, std::string* query, std::vector<ContentValue*> args){

	_db = db;
	_stmt = NULL;
	_sql = query;

	if (!_db->prepareSql(query, &_stmt)){
		IStringBuilder* errBuilder = IStringBuilder::newStringBuilder();
		errBuilder->addString("Can't prepare statement: %s, error=[");
		errBuilder->addInt(_db->errorCode());
		errBuilder->addString(": ");
		errBuilder->addString(*_db->errorMessage());

		std::string error = errBuilder->getString();
		delete errBuilder;

		throw std::exception(error.c_str());
	}

	_queryParamCount = sqlite3_bind_parameter_count(_stmt);
	ILogger::instance()->logInfo("Query param count: %d", _queryParamCount);

	int i = 0;
	while (i++ < _queryParamCount) {
		_db->bindObjectAtColumnToStatement(args.at(i - 1), i, _stmt);
	}

	_columnCount = sqlite3_column_count(_stmt);
	ILogger::instance()->logInfo("Column count: %d", _columnCount);
}

bool SQResultSet::next() const{
	bool result = _db->hasData(_stmt);
	if (!result) {
		this->close();
	}
	return result;
}

void SQResultSet::close() const{
	if (_stmt) {
		sqlite3_finalize(_stmt);
		_stmt = NULL;
	}
}

int SQResultSet::integerColumnByIndex(int index) const{
	int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return 0;
	}

	return sqlite3_column_int(_stmt, index);
}

std::string* SQResultSet::stringColumnByIndex(int index) const{
	int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return NULL;
	}

	const char* text = (const char*)sqlite3_column_text(_stmt, index);
	std::string* str = new std::string(text);// str(text);

	return str;
}

const unsigned char* SQResultSet::dataColumnByIndex(int index, int &length) const{
	int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return NULL;
	}
	length = sqlite3_column_bytes(_stmt, index);
	return (const unsigned char*)sqlite3_column_blob(_stmt, index);

	//return[NSData dataWithBytes : sqlite3_column_blob(_stmt, index)
	//length : (NSUInteger)blobLength];
}



double SQResultSet::doubleColumnByIndex(int index) const{
	int columnType = sqlite3_column_type(_stmt, index);

	if (columnType == SQLITE_NULL) {
		return 0;
	}

	return sqlite3_column_double(_stmt, index);
}

bool SQResultSet::isNullColumnByIndex(int index) const{
	int columnType = sqlite3_column_type(_stmt, index);
	return (columnType == SQLITE_NULL);
}


//----------------------------------------------------------------------------------




SQDatabase::SQDatabase(const std::string dbPath){

		_dbPath = dbPath;
		_busyRetryTimeout = 10;
		_db = NULL;
}

SQDatabase* SQDatabase::initWithPath(const std::string dbPath) {

	try{
		return new SQDatabase(dbPath);
	}
	catch (std::string e){
		ILogger::instance()->logError(e.c_str());
	}
	return NULL;
}

bool SQDatabase::isOpen() const{
	if (_db != NULL)
		return true;
	else
		return false;
}

bool SQDatabase::openReadWrite() const{
	this->close();

	const char* dbpathC = _dbPath.c_str();

	if (sqlite3_open_v2(dbpathC, &_db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, NULL) == SQLITE_OK) {
		return true;
	}
	else {
		_db = NULL;
		ILogger::instance()->logError("Can't open readwrite database: %s", dbpathC);
		return false;
	}
}

bool SQDatabase::openReadOnly() const{
	this->close();

	const char* dbpathC = _dbPath.c_str();

	if (sqlite3_open_v2(dbpathC, &_db, SQLITE_OPEN_READONLY, NULL) == SQLITE_OK) {
		return true;
	}
	else {
		_db = NULL;
		ILogger::instance()->logError("Can't open readonly database: %s", dbpathC);
		return false;
	}
}

void SQDatabase::close() const{
	if (this->isOpen()) {
		sqlite3_close(_db);
		_db = NULL;
	}
}

SQResultSet* SQDatabase::executeQuery(std::string* sql) const{
	
	//va_list args;
	//va_start(args, sql);

	//std::vector<unsigned char*> argsArray;
	//
	//for (int i = 0; i < sql->length(); ++i) {
	//	if (sql->at(i) == '?'){
	//		argsArray.push_back(va_arg(args, unsigned char*)); //TODO: ??
	//	}
	//}

	//va_end(args);

	std::vector<ContentValue*> argsArray = std::vector<ContentValue*>();

	return this->executeQuery(sql, argsArray);
}

SQResultSet* SQDatabase::executeQuery(std::string* sql, std::vector<ContentValue*> args) const{

	return SQResultSet::initForDatabase(this, sql, args);
}

bool SQDatabase::prepareSql(std::string* sql, sqlite3_stmt** stmt) const{
	int numOfRetries = 0;

	do {
		int rc = sqlite3_prepare_v2(_db, sql->c_str(), -1, stmt, NULL);
		if (rc == SQLITE_OK) {
			return true;
		}

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				ILogger::instance()->logError("SQLite Busy 1: %s", _dbPath.c_str());
				break;
			}

			std::this_thread::sleep_for(std::chrono::milliseconds(10) /* 10ms */);
		}
		else {
			ILogger::instance()->logError("SQLite Prepare Failed: %s", sqlite3_errmsg(_db));
			ILogger::instance()->logError(" - Query: %s", sql->c_str());
			break;
		}
	} while (numOfRetries++ <= _busyRetryTimeout);

	return false;
}

bool SQDatabase::hasData(sqlite3_stmt* stmt) const{

	int numOfRetries = 0;

	do {
		int rc = sqlite3_step(stmt);

		if (rc == SQLITE_ROW) {
			return true;
		}

		if (rc == SQLITE_DONE) {
			break;
		}

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				ILogger::instance()->logError("SQLite Busy 2: %s", _dbPath.c_str());
				break;
			}

			std::this_thread::sleep_for(std::chrono::milliseconds(10) /* 10ms */);
		}
		else {
			ILogger::instance()->logError("SQLite Prepare Failed: %s", sqlite3_errmsg(_db));
			break;
		}
	} while (numOfRetries++ <= _busyRetryTimeout);

	return false;
}

int SQDatabase::errorCode() const{
	return sqlite3_errcode(_db);
}

const std::string* SQDatabase::errorMessage() const{
	const char* text = sqlite3_errmsg(_db);
	std::string str(text);

	return &str;
}

double getTimeStamp(std::time_t* t) {
	std::stringstream ts;
	ts << *t;
	std::string timeStamp = ts.str();

	return atof(timeStamp.c_str());
}

void SQDatabase::bindObjectAtColumnToStatement(ContentValue* obj, int idx, sqlite3_stmt* stmt) const{

	
	if (obj == NULL) {
		sqlite3_bind_null(stmt, idx);
	}
	else if (obj->_type == "string"){
		std::string* str = (std::string*)(obj->_content);
		//const char * cstr = str->c_str();
		//std::string* str = new std::string((const char*)(obj->_content));
		//const char* c = (const char*)(obj->_content);
		ILogger::instance()->logInfo("Typeid del OBJ: %s", obj->_type.c_str());
		ILogger::instance()->logInfo("Content del OBJ: %s", str->c_str());
		sqlite3_bind_text(stmt, idx, str->c_str(), -1, SQLITE_STATIC);
	}
	else {
		//char* dataArr = (char*)(obj->_content);
		//int length = strlen(dataArr);
		ILogger::instance()->logInfo("Typeid del OBJ: %s", obj->_type.c_str());
		sqlite3_bind_blob(stmt, idx, obj->_content, obj->_size, SQLITE_STATIC);
	}
	//else if (obj->_type == "unsigned char"){
	//	
	//	if (typeid(obj->_content).name() == "time_t"){
	//		sqlite3_bind_double(stmt, idx, getTimeStamp((time_t*)(obj->_content)));
	//	}
	//	else if (typeid(obj->_content).name() == "bool"){
	//		sqlite3_bind_int(stmt, idx, *((bool*)(obj->_content)) ? 1 : 0);
	//	}
	//	else if (typeid(obj->_content).name() == "int"){
	//		sqlite3_bind_int64(stmt, idx, (long)*((int*)(obj->_content)));
	//	}
	//	else if (typeid(obj->_content).name() == "long"){
	//		sqlite3_bind_int64(stmt, idx, *((long*)(obj->_content)));
	//	}
	//	else if (typeid(obj->_content).name() == "float"){
	//		sqlite3_bind_double(stmt, idx, (double)*((float*)(obj->_content)));
	//	}
	//	else if (typeid(obj->_content).name() == "double"){
	//		sqlite3_bind_double(stmt, idx, *((double*)(obj->_content)));
	//	}
	//	else {
	//		char* dataArr = (char*)(obj->_content);
	//		//int length = strlen(dataArr);
	//		sqlite3_bind_blob(stmt, idx, obj->_content, strlen(dataArr), SQLITE_STATIC);
	//	}
	//}
}

bool SQDatabase::executeStatament(sqlite3_stmt * stmt) const{
	int numOfRetries = 0;
	int rc;

	do {
		rc = sqlite3_step(stmt);
		if (rc == SQLITE_OK || rc == SQLITE_DONE)
			return true;

		if (rc == SQLITE_BUSY) {
			if (numOfRetries == _busyRetryTimeout) {
				ILogger::instance()->logError("SQLite Busy 3: %s", _dbPath.c_str());
				break;
			}

			std::this_thread::sleep_for(std::chrono::milliseconds(10) /* 10ms */);
		}
		else {
			ILogger::instance()->logError("SQLite Step Failed: %s", sqlite3_errmsg(_db));
			break;
		}
	} while (numOfRetries++ <= _busyRetryTimeout);

	return false;
}

bool SQDatabase::executeNonQuery(std::string* sql) const{

	//va_list args;
	//va_start(args, sql);

	//std::vector<unsigned char*> argsArray;
	////argsArray.reserve(3);
	//ILogger::instance()->logInfo("EXECUTENONQUERY SQL= %s", sql->c_str());
	//for (int i = 0; i < sql->length(); ++i) {
	//	if (sql->at(i) == '?'){
	//		ILogger::instance()->logInfo("Typeid del OBJ: %s", typeid(args).name());
	//		argsArray.push_back(va_arg(args, unsigned char*)); //TODO: works??
	//	}
	//}

	//va_end(args);
	ILogger::instance()->logInfo("EXECUTENONQUERY SQL= %s", sql->c_str());
	std::vector<ContentValue*> argsArray = std::vector<ContentValue*>(); //empty args

	return this->executeNonQuery(sql, argsArray);
}

bool SQDatabase::executeNonQuery(std::string* sql, std::vector<ContentValue*> args) const{

	sqlite3_stmt *sqlStmt;

	if (!this->prepareSql(sql, (&sqlStmt))){
		return false;
	}

	int i = 0;
	int queryParamCount = sqlite3_bind_parameter_count(sqlStmt);
	while (i++ < queryParamCount) {
		this->bindObjectAtColumnToStatement(args.at(i - 1), i, sqlStmt);
	}

	bool success = this->executeStatament(sqlStmt);

	sqlite3_finalize(sqlStmt);
	return success;
}

bool SQDatabase::beginTransaction() const{
	std::string str("BEGIN EXCLUSIVE TRANSACTION;");
	return this->executeNonQuery(&str);
}

bool SQDatabase::commit() const{
	std::string str("COMMIT TRANSACTION;");
	return this->executeNonQuery(&str);
}

bool SQDatabase::rollback() const{
	std::string str("ROLLBACK TRANSACTION;");
	return this->executeNonQuery(&str);
}

bool SQDatabase::beginDeferredTransaction() const{
	std::string str("BEGIN DEFERRED TRANSACTION;");
	return this->executeNonQuery(&str);
}

SQDatabase::~SQDatabase(){
	//TODO: 
}
